llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18235,6 +18235,16 @@ SDValue DAGCombiner::combineRepeatedFPDivisors(SDNode *N) {
   if (N0CFP && (N0CFP->isExactlyValue(1.0) || N0CFP->isExactlyValue(-1.0)))
     return SDValue();
 
+  // Skip if we have subnormals, multiplying with the reciprocal will introduce
+  // infinities.
+  ConstantFPSDNode *N1CFP = isConstOrConstSplatFP(N1, /* AllowUndefs */ true);
+  if (N1CFP) {
```

Comment:
You might also want to block the transformation in cases where N1CFP is a non-splat constant vector and one of the elements is subnormal. 

https://godbolt.org/z/zzPT4dTP5